juju/juju
# [JUJU-1214] Remove Windows charm support from Juju for 3.0

Remove Windows support from Juju as an agent.

With the release of Juju 3.0, Juju will know longer support deploying
windows based charms. This commit does the work to remove all Windows
references and code from the code base related to deploying Windows
charms.

Juju will still support Windows for the client and as such some Windows
references have left behind as the client has references to agent based
code.

Where appropriate all tests have been kept or altered to another series
such that test coverage is not affected.

## QA steps

The Windows client should still compile:
```
GOOS=windows go build -a github.com/juju/juju/cmd/juju
```
All the CI below should pass as well (except the ones with known failures).

We can't add Windows machines:
```console
$ juju add-machine --series win7
$ juju status
...
Machine  State  Inst id  Series  AZ  Message
...
1        down   pending  win7        unknown OS for series: "win7"
...
```

## To keep

To support the Windows client, (I believe) we need to keep Windows stuff in:
- `caas/kubernetes/provider/exec/termsize_windows.go`
- `cmd/plugins`
- `core/os`
- `core/paths`
- `juju/osenv`
- `juju/sockets`
- `service/systemd`
- `state/backups`
- `wrench`

We should also keep historical version numbers containing Windows (e.g. `1.2.3-windows-alpha`)

## Changes

- Deleted:
  - `acceptancetests/repository/charms-win/`
  - (list to be completed)
- Removed all Windows stuff in:
  - `apiserver/`
  - `cloudconfig/`
  - Several tests in `cmd/juju/commands/`
  - `cmd/juju/machine/add.go`
  - `cmd/juju/romulus/agree/agree_test.go`
  - Some tests in `cmd/juju/ssh/`
  - `cmd/jujuc/`
  - `cmd/jujud/`
  - `cmd/containeragent/`
  - `provider/`
  - `state/`
  - `worker/`
  - (list to be completed)
- Changed the install targets in Makefile so that on Windows, it will only try to install the client.
- Removed agent binary names from `juju/names/names_windows`

## Questions

I'm not sure if I should:
- Remove Windows from:
  - `agent/` - is this used by the client?
  - `caas/kubernetes/provider/exec/termsize_windows.go`
  - `environs/tools/simplestreams_test.go`
